File: backend/daemon-ts/src/tools/web.ts
/**
 * Web tools — `web_search` (Tavily) and `fetch_url` (HTTP GET + HTML strip).
 *
 * Ported from `backend/daemon/src/tools/web.rs`. Both use Bun's global
 * `fetch` with a 30 s `AbortSignal.timeout`. `fetch_url` strips HTML
 * server-side and truncates to 50 000 characters; the HTML stripper is
 * the careful unicode-safe version (mirrors strip_html in web.rs:199-280).
 */

import type { ToolContext, ToolHandler } from "./registry.ts";
import { ToolError } from "./registry.ts";

export const WEB_SEARCH_DESCRIPTION =
  "Search the web and return a list of results with titles, URLs, and content snippets. Use it to gather more info about a topic, when you're uncertain about a fact, or whenever the conversation touches on recent events, specific people, products, or things that may have happened after your training cutoff. Don't hedge or caveat about whether you 'should' look something up — searching is free, so if it would help, just do it. Pair with `fetch_url` to read the full content of a result page when a snippet isn't enough.";

export const FETCH_URL_DESCRIPTION =
  "Fetch and read the content of a web page, returning its text. Use when a `web_search` result looks relevant and you need the full context, when {{user}} pastes or mentions a URL you should actually read, or when you want to follow a specific page you already have in mind. Best paired with `web_search` to find candidate URLs first. Returns plain text; complex interactive pages, paywalled content, or heavily JS-rendered sites may return limited or no content.";

const MAX_CONTENT_CHARS = 50_000;
const REQUEST_TIMEOUT_MS = 30_000;

// ---------------------------------------------------------------------------
// web_search
// ---------------------------------------------------------------------------

export const webSearchHandler: ToolHandler = {
  name: "web_search",
  description: WEB_SEARCH_DESCRIPTION,
  inputSchema: {
    type: "object",
    properties: {
      query: { type: "string", description: "The search query." },
      max_results: {
        type: "integer",
        description: "Maximum number of results to return.",
        default: 5,
      },
    },
    required: ["query"],
  },
  async execute(input: unknown, ctx: ToolContext): Promise<string> {
    const obj = (input ?? {}) as Record<string, unknown>;
    const query = obj["query"];
    if (typeof query !== "string" || query.length === 0) {
      throw new ToolError("InvalidArgs", "missing 'query' field");
    }
    const cfg = ctx.searchConfig;
    const apiKey = process.env[cfg.api_key_env];
    if (apiKey === undefined || apiKey.length === 0) {
      throw new ToolError(
        "InvalidArgs",
        `web_search requires the ${cfg.api_key_env} environment variable to be set`,
      );
    }
    const maxResults =
      typeof obj["max_results"] === "number" && Number.isFinite(obj["max_results"])
        ? Math.max(1, Math.floor(obj["max_results"] as number))
        : cfg.max_results;

    let res: Response;
    try {
      res = await fetch("https://api.tavily.com/search", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          api_key: apiKey,
          query,
          max_results: maxResults,
          search_depth: cfg.search_depth,
          include_answer: cfg.include_answer,
        }),
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
    } catch (e) {
      throw new ToolError(
        "Http",
        `Tavily request failed: ${(e as Error).message}`,
      );
    }

    if (!res.ok) {
      const errBody = await res.text().catch(() => "");
      throw new ToolError(
        "Http",
        `Tavily API returned HTTP ${res.status}: ${errBody}`,
      );
    }

    let body: unknown;
    try {
      body = await res.json();
    } catch (e) {
      throw new ToolError(
        "Http",
        `failed to parse Tavily response: ${(e as Error).message}`,
      );
    }

    const rawResults = (body as { results?: unknown }).results;
    const results = Array.isArray(rawResults)
      ? rawResults.map((r) => {
          const rec = (r ?? {}) as Record<string, unknown>;
          return {
            title: typeof rec["title"] === "string" ? rec["title"] : "",
            url: typeof rec["url"] === "string" ? rec["url"] : "",
            content: typeof rec["content"] === "string" ? rec["content"] : "",
          };
        })
      : [];

    const out: Record<string, unknown> = { query, results };
    const answer = (body as { answer?: unknown }).answer;
    if (typeof answer === "string") out["answer"] = answer;
    return JSON.stringify(out);
  },
};

// ---------------------------------------------------------------------------
// fetch_url
// ---------------------------------------------------------------------------

export const fetchUrlHandler: ToolHandler = {
  name: "fetch_url",
  description: FETCH_URL_DESCRIPTION,
  inputSchema: {
    type: "object",
    properties: {
      url: { type: "string", description: "The URL to fetch." },
    },
    required: ["url"],
  },
  async execute(input: unknown): Promise<string> {
    const obj = (input ?? {}) as Record<string, unknown>;
    const url = obj["url"];
    if (typeof url !== "string" || url.length === 0) {
      throw new ToolError("InvalidArgs", "missing 'url' field");
    }

    let res: Response;
    try {
      res = await fetch(url, {
        method: "GET",
        redirect: "follow",
        headers: { "User-Agent": "shore/2.0" },
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
    } catch (e) {
      throw new ToolError("Http", `request failed: ${(e as Error).message}`);
    }
    if (!res.ok) {
      throw new ToolError("Http", `HTTP ${res.status} for ${url}`);
    }

    const contentType = res.headers.get("content-type") ?? "unknown";
    let body: string;
    try {
      body = await res.text();
    } catch (e) {
      throw new ToolError(
        "Http",
        `failed to read body: ${(e as Error).message}`,
      );
    }

    const isHtml = contentType.includes("html");
    let content = isHtml ? stripHtml(body) : body;

    const truncated = content.length > MAX_CONTENT_CHARS;
    if (truncated) {
      // Char-boundary safe truncation: substring() on UTF-16 code units
      // is fine for this purpose; we don't need byte-exact behavior.
      content = content.slice(0, MAX_CONTENT_CHARS);
    }

    return JSON.stringify({
      url,
      content_type: contentType,
      content,
      truncated,
    });
  },
};

/**
 * Strip HTML tags and extract readable text content.
 *
 * Phases mirror `strip_html` in web.rs:199-280:
 *   1. Remove `<script>`, `<style>`, `<head>` blocks entirely
 *   2. Strip remaining tags (each tag becomes a single space)
 *   3. Decode common HTML entities
 *   4. Collapse whitespace
 *
 * The case-insensitive tag detection compares ASCII-lowercased slices
 * of the *original* string against the tag list, so byte offsets stay
 * valid even when characters like ẞ → ß would shrink under Unicode
 * lowercasing. The TS port doesn't have the same byte-offset risk
 * (we work on the JS string directly), but we preserve the algorithmic
 * shape for parity.
 */
export function stripHtml(html: string): string {
  const skipTags = ["script", "style", "head"];
  let cleaned = "";
  let i = 0;

  while (i < html.length) {
    if (html[i] === "<") {
      const remainingLower = html.slice(i).toLowerCase();
      const skip = skipTags.find((t) => remainingLower.startsWith(`<${t}`));
      if (skip !== undefined) {
        const closeMarker = `</${skip}`;
        const endPos = remainingLower.indexOf(closeMarker);
        if (endPos >= 0) {
          const afterClose = i + endPos + closeMarker.length;
          const gt = html.indexOf(">", afterClose);
          if (gt >= 0) {
            i = gt + 1;
            continue;
          }
        }
        break;
      }
      const gt = html.indexOf(">", i);
      if (gt >= 0) {
        cleaned += " ";
        i = gt + 1;
        continue;
      }
    }
    cleaned += html[i];
    i += 1;
  }

  // Phase 2: Decode common HTML entities (verbatim from web.rs:253-262).
  const decoded = cleaned
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'")
    .replace(/&nbsp;/g, " ")
    .replace(/&#x27;/g, "'")
    .replace(/&#x2F;/g, "/");

  // Phase 3: Collapse whitespace.
  let result = "";
  let prevWs = false;
  for (const ch of decoded) {
    if (/\s/.test(ch)) {
      if (!prevWs) result += " ";
      prevWs = true;
    } else {
      result += ch;
      prevWs = false;
    }
  }
  return result.trim();
}

File: backend/daemon/src/tools/web.rs
use super::{ToolCategory, ToolContext, ToolDef, ToolError};
use serde_json::{json, Value};
use std::time::Duration;

// ---------------------------------------------------------------------------
// Tool definitions
// ---------------------------------------------------------------------------

pub fn tool_defs() -> Vec<ToolDef> {
    vec![
        ToolDef {
            name: "web_search",
            description: crate::include_prompt!("../../prompts/tools/web/web_search.md"),
            parameters: json!({
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "The search query."
                    },
                    "max_results": {
                        "type": "integer",
                        "description": "Maximum number of results to return.",
                        "default": 5
                    }
                },
                "required": ["query"]
            }),
            category: ToolCategory::Web,
        },
        ToolDef {
            name: "fetch_url",
            description: crate::include_prompt!("../../prompts/tools/web/fetch_url.md"),
            parameters: json!({
                "type": "object",
                "properties": {
                    "url": {
                        "type": "string",
                        "description": "The URL to fetch."
                    }
                },
                "required": ["url"]
            }),
            category: ToolCategory::Web,
        },
    ]
}

// ---------------------------------------------------------------------------
// Handlers
// ---------------------------------------------------------------------------

/// Handle `web_search` — search the web via Tavily API.
pub async fn handle_web_search(input: Value, ctx: &dyn ToolContext) -> Result<Value, ToolError> {
    let query = input
        .get("query")
        .and_then(|v| v.as_str())
        .ok_or_else(|| ToolError::InvalidArgs("missing 'query' field".to_string()))?;

    let search_cfg = ctx.search_config();

    let api_key = std::env::var(&search_cfg.api_key_env).map_err(|_| {
        ToolError::InvalidArgs(format!(
            "web_search requires the {} environment variable to be set",
            search_cfg.api_key_env
        ))
    })?;

    let max_results = input
        .get("max_results")
        .and_then(Value::as_u64)
        .unwrap_or(u64::from(search_cfg.max_results));

    let body = json!({
        "api_key": api_key,
        "query": query,
        "max_results": max_results,
        "search_depth": search_cfg.search_depth,
        "include_answer": search_cfg.include_answer,
    });

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(30))
        .build()
        .map_err(|e| ToolError::Http(format!("failed to build HTTP client: {e}")))?;

    let resp = client
        .post("https://api.tavily.com/search")
        .json(&body)
        .send()
        .await
        .map_err(|e| ToolError::Http(format!("Tavily request failed: {e}")))?;

    let status = resp.status().as_u16();
    if !resp.status().is_success() {
        let err_body = resp.text().await.unwrap_or_default();
        return Err(ToolError::Http(format!(
            "Tavily API returned HTTP {status}: {err_body}"
        )));
    }

    let tavily_resp: Value = resp
        .json()
        .await
        .map_err(|e| ToolError::Http(format!("failed to parse Tavily response: {e}")))?;

    // Reshape results to a clean format.
    let results = tavily_resp
        .get("results")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .map(|r| {
                    json!({
                        "title": r.get("title").and_then(|v| v.as_str()).unwrap_or(""),
                        "url": r.get("url").and_then(|v| v.as_str()).unwrap_or(""),
                        "content": r.get("content").and_then(|v| v.as_str()).unwrap_or(""),
                    })
                })
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();

    let mut result = json!({
        "query": query,
        "results": results,
    });

    if let Some(answer) = tavily_resp.get("answer").and_then(|v| v.as_str()) {
        if let Some(obj) = result.as_object_mut() {
            let _ignored = obj.insert("answer".into(), json!(answer));
        }
    }

    Ok(result)
}

/// Maximum content length returned to the model (chars).
const MAX_CONTENT_CHARS: usize = 50_000;

/// Handle `fetch_url` — fetch a webpage and extract readable text.
pub async fn handle_fetch_url(input: Value) -> Result<Value, ToolError> {
    let url = input
        .get("url")
        .and_then(|v| v.as_str())
        .ok_or_else(|| ToolError::InvalidArgs("missing 'url' field".to_string()))?;

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(30))
        .user_agent("shore/2.0")
        .redirect(reqwest::redirect::Policy::limited(5))
        .build()
        .map_err(|e| ToolError::Http(format!("failed to build HTTP client: {e}")))?;

    let resp = client
        .get(url)
        .send()
        .await
        .map_err(|e| ToolError::Http(format!("request failed: {e}")))?;

    let status = resp.status().as_u16();
    if !resp.status().is_success() {
        return Err(ToolError::Http(format!("HTTP {status} for {url}")));
    }

    let content_type = resp
        .headers()
        .get(reqwest::header::CONTENT_TYPE)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("unknown")
        .to_string();

    let body = resp
        .text()
        .await
        .map_err(|e| ToolError::Http(format!("failed to read body: {e}")))?;

    let is_html = content_type.contains("html");
    let content = if is_html { strip_html(&body) } else { body };

    let truncated = content.len() > MAX_CONTENT_CHARS;
    let content = if truncated {
        let boundary = content.floor_char_boundary(MAX_CONTENT_CHARS);
        content[..boundary].to_string()
    } else {
        content
    };

    Ok(json!({
        "url": url,
        "content_type": content_type,
        "content": content,
        "truncated": truncated,
    }))
}

/// Strip HTML tags and extract readable text content.
///
/// Removes `<script>`, `<style>`, and `<head>` blocks entirely, strips remaining
/// tags, decodes common HTML entities, and collapses whitespace.
fn strip_html(html: &str) -> String {
    // Phase 1: Remove script, style, and head blocks.
    //
    // We work entirely on the original `html` string using byte offsets
    // that are always derived from `html` itself.  Case-insensitive
    // comparisons use `str::to_ascii_lowercase()` on small slices rather
    // than a parallel lowercased copy (which can differ in byte length
    // for certain Unicode characters like ẞ → ß).
    let mut cleaned = String::with_capacity(html.len());
    let bytes = html.as_bytes();
    let mut i = 0;

    while i < html.len() {
        if bytes.get(i) == Some(&b'<') {
            // Check for blocks we want to skip entirely.
            // Compare case-insensitively against the original html slice.
            let remaining_lower = html[i..].to_ascii_lowercase();
            if let Some(tag) = ["script", "style", "head"]
                .iter()
                .find(|tag| remaining_lower.starts_with(&format!("<{tag}")))
            {
                // Find the closing tag (case-insensitive) in the original string.
                let close = format!("</{tag}");
                let rest_lower = html[i..].to_ascii_lowercase();
                if let Some(end_pos) = rest_lower.find(&close) {
                    // Skip past the closing tag's '>'
                    let after_close = i.saturating_add(end_pos).saturating_add(close.len());
                    if let Some(gt) = html[after_close..].find('>') {
                        i = after_close.saturating_add(gt).saturating_add(1);
                        continue;
                    }
                }
                // No closing tag found — skip to end
                break;
            }

            // Regular tag — skip it but add a space (block elements create breaks)
            if let Some(gt) = html[i..].find('>') {
                cleaned.push(' ');
                i = i.saturating_add(gt).saturating_add(1);
                continue;
            }
        }

        // Consume one character
        if let Some(ch) = html[i..].chars().next() {
            cleaned.push(ch);
            i = i.saturating_add(ch.len_utf8());
        } else {
            break;
        }
    }

    // Phase 2: Decode common HTML entities
    let decoded = cleaned
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&#39;", "'")
        .replace("&apos;", "'")
        .replace("&nbsp;", " ")
        .replace("&#x27;", "'")
        .replace("&#x2F;", "/");

    // Phase 3: Collapse whitespace
    let mut result = String::with_capacity(decoded.len());
    let mut prev_whitespace = false;
    for ch in decoded.chars() {
        if ch.is_whitespace() {
            if !prev_whitespace {
                result.push(' ');
            }
            prev_whitespace = true;
        } else {
            result.push(ch);
            prev_whitespace = false;
        }
    }

    result.trim().to_string()
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::TestToolContext;

    #[test]
    fn test_web_tool_defs() {
        let defs = tool_defs();
        assert_eq!(defs.len(), 2);

        let names: Vec<&str> = defs.iter().map(|d| d.name).collect();
        assert!(names.contains(&"web_search"));
        assert!(names.contains(&"fetch_url"));

        // All web tools should have Web category.
        for def in &defs {
            assert_eq!(def.category, ToolCategory::Web);
        }
    }

    #[tokio::test]
    async fn test_web_search_missing_query() {
        let ctx = TestToolContext::new();
        let result = handle_web_search(json!({}), &ctx).await;
        assert!(matches!(result, Err(ToolError::InvalidArgs(_))));
    }

    #[tokio::test]
    async fn test_web_search_no_api_key() {
        // With default config, TAVILY_API_KEY is unlikely to be set in test env.
        let ctx = TestToolContext::new();
        let result = handle_web_search(json!({"query": "rust programming"}), &ctx).await;
        assert!(matches!(result, Err(ToolError::InvalidArgs(_))));
        if let Err(ToolError::InvalidArgs(msg)) = result {
            assert!(
                msg.contains("TAVILY_API_KEY"),
                "error should name the env var: {msg}"
            );
        }
    }

    /// Live integration test — requires TAVILY_API_KEY env var.
    #[tokio::test]
    #[ignore = "live web search requires TAVILY_API_KEY and network access"]
    async fn test_web_search_live() {
        let ctx = TestToolContext::new();
        let result = handle_web_search(
            json!({"query": "Rust programming language", "max_results": 2}),
            &ctx,
        )
        .await
        .expect("live search should succeed");

        assert_eq!(result["query"], "Rust programming language");
        let results = result["results"]
            .as_array()
            .expect("results should be array");
        assert!(!results.is_empty(), "should have at least one result");
        // Each result should have title, url, content.
        for r in results {
            assert!(r["title"].as_str().is_some());
            assert!(r["url"].as_str().is_some());
            assert!(r["content"].as_str().is_some());
        }
    }

    #[tokio::test]
    async fn test_fetch_url_missing_url() {
        let result = handle_fetch_url(json!({})).await;
        assert!(matches!(result, Err(ToolError::InvalidArgs(_))));
    }

    #[test]
    fn test_strip_html_basic() {
        let html = "<html><body><h1>Hello</h1><p>World</p></body></html>";
        let text = strip_html(html);
        assert!(text.contains("Hello"));
        assert!(text.contains("World"));
        assert!(!text.contains("<h1>"));
    }

    #[test]
    fn test_strip_html_removes_script_and_style() {
        let html = r"<html><head><title>T</title></head><body>
            <script>var x = 1;</script>
            <style>.foo { color: red; }</style>
            <p>Visible text</p>
        </body></html>";
        let text = strip_html(html);
        assert!(text.contains("Visible text"));
        assert!(!text.contains("var x"));
        assert!(!text.contains("color: red"));
        assert!(!text.contains("<title>"));
    }

    #[test]
    fn test_strip_html_decodes_entities() {
        let html = "<p>A &amp; B &lt; C &gt; D &quot;E&quot;</p>";
        let text = strip_html(html);
        assert!(text.contains("A & B < C > D \"E\""));
    }

    #[test]
    fn test_strip_html_collapses_whitespace() {
        let html = "<p>  lots   of    spaces  </p>";
        let text = strip_html(html);
        // Should not have runs of multiple spaces
        assert!(!text.contains("  "));
    }

    /// `ẞ` (U+1E9E, 3 bytes) lowercases to `ß` (U+00DF, 2 bytes).
    /// When `to_lowercase()` shrinks the string, byte offsets from `html`
    /// exceed `lower_bytes.len()` → out-of-bounds panic.
    #[test]
    fn test_strip_html_unicode_lowercase_length_change() {
        // ẞ (3 bytes) → ß (2 bytes). After 10 ẞ chars:
        //   html  = 30 bytes of ẞ + 10 bytes of tag = 40 bytes
        //   lower = 20 bytes of ß + 10 bytes of tag = 30 bytes
        // At offset 30 in html, lower_bytes[30] is out of bounds.
        let html = format!("{}<b>x</b>", "ẞ".repeat(10));
        let result = strip_html(&html);
        // Should extract the text content without panicking.
        assert!(result.contains('x'));
    }
}
